### ISSUE TITLE
`get_first` doesn't lint on slices of non-primitive types

### ISSUE THREAD

**Issue by y21** [rust-lang/rust-clippy#11594]
### Summary

We have a lint for `slice.get(0)`, but I've noticed that in many cases this just doesn't warn.  I looked into the code and there's this check:
https://github.com/rust-lang/rust-clippy/blob/aee3daf90ba36fb14e3b9f04ea0eea97fd2ec6ad/clippy_lints/src/methods/get_first.rs#L23

I'm honestly not sure why this is there. `.first()` should work just as well with non-primitives?
I removed the check locally and all tests still pass so it remains a mystery to me. Doesn't look like that was mentioned in the initial PR or in any issues. Opening an issue first because someone might have a clue as to why it's necessary.

### Lint Name

`get_first`

### Reproducer

I tried this code:

```rust
let strings = &[String::new()];
let v: Option<&String> = strings.get(0);
```

I expected to see this happen: 
```rs
warning: accessing first element with `strings.get(0)`
   --> test.rs:164:30
    |
164 |     let v: Option<&String> = strings.get(0);
    |                              ^^^^^^^^^^^^^^ help: try: `strings.first()`
```

Instead, this happened:
nothing.

### Version

```text
rustc 1.74.0-nightly (37390d656 2023-09-24)
binary: rustc
commit-hash: 37390d65636dd67e263753a3c04fbc60dcc4348e
commit-date: 2023-09-24
host: x86_64-unknown-linux-gnu
release: 1.74.0-nightly
LLVM version: 17.0.0
```
